README.md
# Yall

<p align="center">
    <img alt="Yall logo"  src="https://github.com/Amber-Williams/yall/blob/main/quick-draw.gif" width="500"/>
</p>
<p align="center">
  <a href="https://github.com/Amber-Williams/yall/actions/workflows/release.yml">
    <img alt="Build states" src="https://github.com/Amber-Williams/yall/actions/workflows/release.yml/badge.svg">
  </a>
  <a href="#badge">
    <img alt="semantic-release: angular" src="https://img.shields.io/badge/semantic--release-angular-e10079?logo=semantic-release">
  </a>
  <a href="https://www.npmjs.com/package/yall-cli">
    <img alt="Downloads a month" src="https://img.shields.io/npm/dm/yall-cli.svg">
  </a>
</p>

A CLI tool to launch applications in half the time

<img alt="Welcome to Yall" src="https://github.com/Amber-Williams/yall/blob/main/demo.gif" width="600" />

#### Installation

`npm install -g yall-cli`

Recommended:

- Setup of `code` alias to launch new windows of VSCode [Installation](https://code.visualstudio.com/docs/setup/mac)

#### Customize your flow

Run `yall --edit`

#### Run math evaluations from your terminal

Run `yall --math <eval>`
Examples include:
`yall --math 12,000 + 4,000` => `12400`
`yall --math 12_000 + 4_000` => `12400`
`yall --math 12.7 cm to inch` => `5 inch`
`yall --math sin(45 deg) ^ 2` => `0.5`
`yall --math 9 / 3 + 2i` => `3 + 2i`

## Support

| Operating System    | Support       |
| ------------------- | ------------- |
| macOS (Silicon M1)  | Supported     |
| macOS (Intel chips) | Unknown       |
| Windows             | Not supported |
| Linux               | Not supported |
| Other OS            | Not supported |

</br></br>

| Application        | Support   |
| ------------------ | --------- |
| Chrome             | Supported |
| Visual Studio Code | Supported |
| PyCharm            | Supported |
| Notion             | Supported |


CHANGELOG.md
## [2.3.1](https://github.com/Amber-Williams/yall/compare/v2.3.0...v2.3.1) (2023-08-04)


### Bug Fixes

* bell sound path ([4f10466](https://github.com/Amber-Williams/yall/commit/4f10466ff30b3ba1cf62d9ed4eb377ca343a6cbd))

# [2.3.0](https://github.com/Amber-Williams/yall/compare/v2.2.0...v2.3.0) (2023-08-04)


### Features

* add timer option ([bd5c6b2](https://github.com/Amber-Williams/yall/commit/bd5c6b296ce4cbdd567534be14709ad6720d5683))

# [2.2.0](https://github.com/Amber-Williams/yall/compare/v2.1.0...v2.2.0) (2023-05-04)


### Features

* **cli:** Returns numeric seprators when used in math expression ([8ad694b](https://github.com/Amber-Williams/yall/commit/8ad694b8e2171c76e4b4fcf6f62a94229b1b8697))
* **tape:** updates demo tape ([38a4da4](https://github.com/Amber-Williams/yall/commit/38a4da4934af1d44cb329a9f6b2cfb7c4eb8e4f7))

# [2.1.0](https://github.com/Amber-Williams/yall/compare/v2.0.0...v2.1.0) (2023-05-03)


### Features

* **demo:** Add math commands ([c7e3bed](https://github.com/Amber-Williams/yall/commit/c7e3bed57a4d67c0b8922d6ca218255c826136ef))

# [2.0.0](https://github.com/Amber-Williams/yall/compare/v1.0.0...v2.0.0) (2023-05-03)


### Features

* **cli:** math evaluation ([d843c67](https://github.com/Amber-Williams/yall/commit/d843c67409fb8c925f25162eda20fcb01eaed61f))
* **docs:** add monthly downloads badge ([20ceb35](https://github.com/Amber-Williams/yall/commit/20ceb355b913bf49e1e594e09b309b0ef87a7421))
* **types:** add support for continuous logs such as container logs ([88d02c5](https://github.com/Amber-Williams/yall/commit/88d02c5eb42729276baef48e45830d1b52205198))


### Performance Improvements

* **release:** Fixes release bug ([2b5072c](https://github.com/Amber-Williams/yall/commit/2b5072c391ece3a5ba9b263dc5cfd3e03a027b4e))


### BREAKING CHANGES

* **release:** Fixes release bug

# [1.1.0](https://github.com/Amber-Williams/yall/compare/v1.0.0...v1.1.0) (2023-04-12)


### Features

* **docs:** add monthly downloads badge ([20ceb35](https://github.com/Amber-Williams/yall/commit/20ceb355b913bf49e1e594e09b309b0ef87a7421))
* **types:** add support for continuous logs such as container logs ([88d02c5](https://github.com/Amber-Williams/yall/commit/88d02c5eb42729276baef48e45830d1b52205198))

# [1.1.0](https://github.com/Amber-Williams/yall/compare/v1.0.0...v1.1.0) (2023-04-12)


### Features

* **docs:** add monthly downloads badge ([20ceb35](https://github.com/Amber-Williams/yall/commit/20ceb355b913bf49e1e594e09b309b0ef87a7421))

# 1.0.0 (2023-04-09)


### Features

* demo video ([62a714f](https://github.com/Amber-Williams/yall/commit/62a714f433771b13a840678a066a6079728f4a75))


index.js
#!/usr/bin/env node
import fs from "fs";

import path from "path";

import { program } from "commander";

import logger from "./logger.js";
import * as math from "./modes/math.js";
import * as timer from "./modes/timer.js";
import * as utils from "./utils.js";

const { __dirname } = utils.fileDirName(import.meta);
const packageJsonFilePath = path.join(__dirname, "package.json");
const { description, version } = JSON.parse(
  fs.readFileSync(packageJsonFilePath)
);

program
  .version(version)
  .description(description)
  .option("-e, --edit", "Edit your choices")
  .option(math.option, math.optionDef)
  .option(timer.option, timer.optionDef)
  .action(async (cmd) => {
    if (cmd.edit) {
      logger.warning("Opening choices file...");
      try {
        await utils.openFileInEditor(utils.choicesFilePath);
      } catch (error) {
        logger.error(`An error occurred: ${error}`);
      }
    } else if (cmd.math) {
      await math.logic([cmd.math, ...program.args]);
    } else if (cmd.timer) {
      await timer.logic(cmd.timer, program.args[0]);
    } else {
      utils.askChoices();
    }
  });

program.parse(process.argv);


utils.js
import { exec } from "child_process";
import fs from "fs";
import path, { dirname } from "path";
import { fileURLToPath } from "url";

import inquirer from "inquirer";
import autocompletePrompt from "inquirer-autocomplete-prompt";

import logger from "./logger.js";

inquirer.registerPrompt("autocomplete", autocompletePrompt);

const { __dirname } = fileDirName(import.meta);
export const choicesFilePath = path.join(__dirname, "choices.json");
let choices = JSON.parse(fs.readFileSync(choicesFilePath));

export function fileDirName(meta) {
  const __filename = fileURLToPath(meta.url);

  const __dirname = dirname(__filename);

  return { __dirname, __filename };
}

export function openFileInEditor(filePath) {
  return new Promise((resolve, reject) => {
    let command;
    if (process.platform === "darwin") {
      command = `open -a "Visual Studio Code" ${filePath}`;
    } else {
      reject(new Error(`Unsupported platform: ${process.platform}`));
      return;
    }
    exec(command, (error) => {
      if (error) {
        reject(error);
      } else {
        resolve();
      }
    });
  });
}

export function askChoices() {
  inquirer
    .prompt([
      {
        type: "autocomplete",
        name: "_1",
        message: "Where to?",
        source: (_, input) => {
          input = input || "";
          return Promise.resolve(
            Object.keys(choices).filter((project) =>
              project.toLowerCase().startsWith(input.toLowerCase())
            )
          );
        },
      },
      {
        type: "autocomplete",
        name: "_2",
        message: "And then?",
        source: (answersSoFar, input) => {
          if (typeof choices[answersSoFar["_1"]] === "edit") {
          }

          input = input || "";
          const places = Object.keys(choices[answersSoFar["_1"]]);

          return Promise.resolve(
            places.filter((type) =>
              type.toLowerCase().startsWith(input.toLowerCase())
            )
          );
        },
      },
    ])
    .then((answers) => {
      runScript(answers);
    });
}

export function runScript(answers) {
  let command = choices[answers._1][answers._2];
  const isUrl = command.match("https");

  if (isUrl) {
    command = `open -na "Google Chrome" --args --new-window "${command}"`;
  }

  const childProcess = exec(command, (error, stdout, stderr) => {
    if (error) {
      logger.error(`exec error: ${error}`);
      return;
    }
    if (stderr) {
      logger.error(`stderr: ${stderr}`);
      return;
    }

    logger.branded("yeehaw");

    if (stdout) {
      logger.info(`stdout: ${stdout}`);
    }
  });

  childProcess.stdout.on("data", function (chunk) {
    logger.info(`${chunk}`);
  });
}


lib/player.js
import { exec } from "child_process";
import util from "util";

const execPromise = util.promisify(exec);

/* MAC PLAY COMMAND */
const macPlayCommand = (path, volume) => `afplay \"${path}\" -v ${volume}`;

/* WINDOW PLAY COMMANDS */
const addPresentationCore = `Add-Type -AssemblyName presentationCore;`;
const createMediaPlayer = `$player = New-Object system.windows.media.mediaplayer;`;
const loadAudioFile = (path) => `$player.open('${path}');`;
const playAudio = `$player.Play();`;
const stopAudio = `Start-Sleep 1; Start-Sleep -s $player.NaturalDuration.TimeSpan.TotalSeconds;Exit;`;

const windowPlayCommand = (path, volume) =>
  `powershell -c ${addPresentationCore} ${createMediaPlayer} ${loadAudioFile(
    path
  )} $player.Volume = ${volume}; ${playAudio} ${stopAudio}`;

const player = {
  play: async (path, volume = 0.5) => {
    /**
     * Window: mediaplayer's volume is from 0 to 1, default is 0.5
     * Mac: afplay's volume is from 0 to 255, default is 1. However, volume > 2 usually result in distortion.
     * Therefore, it is better to limit the volume on Mac, and set a common scale of 0 to 1 for simplicity
     */
    const volumeAdjustedByOS =
      process.platform === "darwin" ? Math.min(2, volume * 2) : volume;

    const playCommand =
      process.platform === "darwin"
        ? macPlayCommand(path, volumeAdjustedByOS)
        : windowPlayCommand(path, volumeAdjustedByOS);
    try {
      await execPromise(playCommand, { windowsHide: true });
    } catch (err) {
      throw err;
    }
  },
};

export default player;


modes/timer.js
import path from "path";

import player from "./../lib/player.js";
import logger from "./../logger.js";
import * as utils from "./../utils.js";

const { __dirname } = utils.fileDirName(import.meta);
const bellSoundFile = path.join(__dirname, "/../media/bells.mp3");

export const option = "-t, --timer [time unit] [number of units]";
export const optionDef = `Timer mode
\n
time units available:
  s, sec, secs, second, seconds
  m, min, mins, minute, minutes
  h, hr, hrs, hour, hours
  d, day, days
`;

const soundAlarm = () => {
  player
    .play(bellSoundFile, (err) => {
      if (err) console.log(`Could not play sound: ${err}`);
    })
    .then(() => {
      logger.branded("Time's up!");
    });
};

export const logic = async (unit, num) => {
  if (isNaN(Number(num))) {
    logger.error("Please enter number of mins for timer");
    return;
  }

  let time;
  try {
    switch (unit) {
      case "s":
      case "sec":
      case "secs":
      case "second":
      case "seconds":
        time = num * 1000;
        unit = "seconds";
        break;
      case "m":
      case "min":
      case "mins":
      case "minute":
      case "minutes":
        time = num * 1000 * 60;
        unit = "minutes";
        break;
      case "h":
      case "hr":
      case "hrs":
      case "hour":
      case "hours":
        time = num * 1000 * 60 * 60;
        unit = "hours";
        break;
      case "d":
      case "day":
      case "days":
        time = num * 1000 * 60 * 60 * 24;
        unit = "days";
        break;
      default:
        logger.error("Wrong time unit format was provided");
        break;
    }

    setTimeout(soundAlarm, time);
    logger.branded(`Setting timer for ${num} ${unit}...`);
  } catch {
    logger.error("Unable evaluate the provided equation");
  }
};


modes/math.js
import { evaluate } from "mathjs";

import logger from "./../logger.js";

export const option = "-m, --math [expression]";
export const optionDef = `Math mode
\n
• Handles numeric seprators
  12,000 + 4,000     =>    12,400
  12_000 + 4_000     =>    12_400

• Handles unit converstion expressions
  12.7 cm to inch      =>    5 inch

• Complex expressions
  "sin(45 deg) ^ 2"    =>    0.5
  "9 / 3 + 2i"         =>    3 + 2i
  "det([-1, 2; 3, 1])" =>     -7
`;

export const logic = async (args) => {
  let mathResult;

  try {
    let mathString = args.join(" ");
    const hasCommas = mathString.replaceAll(",", "").length < mathString.length;
    const hasUnderscores =
      mathString.replaceAll("_", "").length < mathString.length;
    mathString = mathString.replaceAll(",", "").replaceAll("_", "");
    mathResult = evaluate(mathString);

    if (hasUnderscores) {
      mathResult = mathResult.toLocaleString().replace(/,/g, "_");
    } else if (hasCommas) {
      mathResult = mathResult.toLocaleString();
    }

    logger.branded(`${mathResult}`);
  } catch {
    logger.error("Unable evaluate the provided equation");
  }
};
